File: wallendar-app/app/src/main/java/com/td/wallendar/groupbalance/GroupBalanceAdapter.java
package com.td.wallendar.groupbalance;

import android.annotation.SuppressLint;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.td.wallendar.R;
import com.td.wallendar.models.Debt;

import java.util.ArrayList;
import java.util.List;

public class GroupBalanceAdapter extends RecyclerView.Adapter<GroupBalanceViewHolder> {

    private final List<Debt> dataset;
    private final long loggedUserId;
    private OnGroupSettleUpClickedListener onGroupSettleUpClickedListener;
    private OnGroupRemindClickedListener onGroupRemindClickedListener;

    public GroupBalanceAdapter(final long loggedUserId) {
        this.dataset = new ArrayList<>();
        this.loggedUserId = loggedUserId;
    }

    public void setOnGroupSettleUpClickedListener(OnGroupSettleUpClickedListener listener) {
        this.onGroupSettleUpClickedListener = listener;
    }

    public void setOnGroupRemindClickedListener(OnGroupRemindClickedListener listener) {
        this.onGroupRemindClickedListener = listener;
    }

    @NonNull
    @Override
    public GroupBalanceViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        final View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.row_balances, parent, false);
        return new GroupBalanceViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull GroupBalanceViewHolder holder, int position) {
        holder.bind(dataset.get(position), loggedUserId);
        holder.setOnGroupSettleUpClickedListener(onGroupSettleUpClickedListener);
        holder.setOnGroupRemindClickedListener(onGroupRemindClickedListener);
    }

    @Override
    public int getItemCount() {
        return dataset == null ? 0 : dataset.size();
    }

    @SuppressLint("NotifyDataSetChanged")
    public void setDataset(final List<Debt> newDataset) {
        this.dataset.clear();
        if (newDataset != null) {
            this.dataset.addAll(newDataset);
        }
        notifyDataSetChanged();
    }

    public void removeFromDataset(Debt debt) {
        int position = this.dataset.indexOf(debt);
        this.dataset.remove(position);
        notifyItemRemoved(position);
    }
}

File: wallendar-app/app/src/main/java/com/td/wallendar/groupbalance/GroupBalanceAdapter.kt
package com.td.wallendar.groupbalance

import android.annotation.SuppressLint
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.td.wallendar.R
import com.td.wallendar.models.Debt
import java.util.*

class GroupBalanceAdapter(loggedUserId: Long) : RecyclerView.Adapter<GroupBalanceViewHolder?>() {
    private val dataset: MutableList<Debt>
    private val loggedUserId: Long
    private var onGroupSettleUpClickedListener: OnGroupSettleUpClickedListener? = null
    private var onGroupRemindClickedListener: OnGroupRemindClickedListener? = null
    fun setOnGroupSettleUpClickedListener(listener: OnGroupSettleUpClickedListener?) {
        onGroupSettleUpClickedListener = listener
    }

    fun setOnGroupRemindClickedListener(listener: OnGroupRemindClickedListener?) {
        onGroupRemindClickedListener = listener
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): GroupBalanceViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.row_balances, parent, false)
        return GroupBalanceViewHolder(view)
    }

    override fun onBindViewHolder(holder: GroupBalanceViewHolder, position: Int) {
        holder.bind(dataset[position], loggedUserId)
        holder.setOnGroupSettleUpClickedListener(onGroupSettleUpClickedListener)
        holder.setOnGroupRemindClickedListener(onGroupRemindClickedListener)
    }

    override fun getItemCount(): Int {
        return dataset.size
    }

    @SuppressLint("NotifyDataSetChanged")
    fun setDataset(newDataset: MutableList<Debt>) {
        dataset.clear()
        dataset.addAll(newDataset)
        notifyDataSetChanged()
    }

    fun removeFromDataset(debt: Debt?) {
        val position = dataset.indexOf(debt)
        dataset.removeAt(position)
        notifyItemRemoved(position)
    }

    init {
        dataset = ArrayList()
        this.loggedUserId = loggedUserId
    }
}